src/redisearch_rs/inverted_index/src/lib.rs

```diff
@@ -497,11 +543,147 @@ impl<E: Encoder> InvertedIndex<E> {
+    repair: RepairType,
+}
+
+/// Information about the result of applying a garbage collection scan to the index
+#[derive(Debug, Eq, PartialEq)]
+pub struct GcApplyInfo {
+    /// The number of bytes that were freed
+    bytes_freed: usize,
+
+    /// The number of bytes that were allocated
+    bytes_allocated: usize,
+
+    /// The number of entries that were removed from the index
+    entries_removed: usize,
+
+    /// The number of blocks that were ignored because the index changed since the scan was performed
+    blocks_ignored: usize,
+}
+
 impl<E: Encoder + DecodedBy> InvertedIndex<E> {
     /// Create a new [`IndexReader`] for this inverted index.
     pub fn reader(&self) -> IndexReader<'_, E, E::Decoder> {
         IndexReader::new(self)
     }
+
+    /// Scan the index for blocks that can be garbage collected. A block can be garbage collected
+    /// if any of its records point to documents that no longer exist. The `doc_exist_cb`
+    /// callback is used to check if a document exists. It should return `true` if the document
+    /// exists and `false` otherwise.
+    ///
+    /// This function returns a delta if GC is needed, or `None` if no GC is needed.
+    pub fn scan_gc(
+        &self,
+        doc_exist_cb: fn(doc_id: t_docId) -> bool,
+    ) -> std::io::Result<Option<GcScanDelta>> {
+        let mut results = Vec::new();
+
+        for (i, block) in self.blocks.iter().enumerate() {
+            if block.num_entries == 0 {
+                results.push(BlockGcScanResult {
+                    index: i,
+                    repair: RepairType::Delete,
+                });
+                continue;
+            }
+
+            let encoder = self.encoder.clone();
+
+            let repair = block.repair(doc_exist_cb, encoder)?;
+
+            if let Some(repair) = repair {
+                results.push(BlockGcScanResult { index: i, repair });
+            }
+        }
+
+        if results.is_empty() {
+            Ok(None)
+        } else {
+            Ok(Some(GcScanDelta {
+                last_block_idx: self.blocks.len() - 1,
+                last_block_num_entries: self.blocks.last().map(|b| b.num_entries).unwrap_or(0),
+                deltas: results,
+            }))
+        }
+    }
+
+    /// Apply the deltas of a garbage collection scan to the index. This will modify the index
+    /// by deleting or repairing blocks as needed.
+    pub fn apply_gc(&mut self, delta: GcScanDelta) -> GcApplyInfo {
+        let GcScanDelta {
+            last_block_idx,
+            last_block_num_entries,
+            deltas,
+        } = delta;
+
+        let mut info = GcApplyInfo {
+            bytes_freed: 0,
+            bytes_allocated: 0,
+            entries_removed: 0,
+            blocks_ignored: 0,
+        };
+
+        // We apply the repairs in reverse order so that the indices of the blocks to be repaired
+        // remain valid as we modify the blocks vector.
+        for delta in deltas.into_iter().rev() {
```

Comment:
It feels like we are doing a lot of shifting here, for both removals and replacements.
I wonder if it'd be faster to have an auxiliary `Vec` with the same capacity of the old one, drain the old one and push the right entries into the new one directly.

Reply:
Done in https://github.com/RediSearch/RediSearch/pull/6878/commits/2100d715e2a9a6feaad7f1fcd7a7205b86c65250